Right click layerlist layer

alukach opened the pull request:
In response to the comments on #205, here's a first-shot at adding right-click menu support for layers on the layerlist.

This entailed a bit of a refactor in the way menus currently worked.  Each menu (left sidebar, operations menu, and right-click context-menu) was separated out into separate files and namespaces (`dc.menus.left`, `dc.menus.geo`, and `dc.menus.layerContextMenu`).  A `'layerlist:added'` signal was created to return newly added layers. 

Currently, the right-click context menus only work on a single layer.

![right-click-layer](https://cloud.githubusercontent.com/assets/897290/10719286/a25db600-7b4b-11e5-9a6f-63ad6b60921d.gif)

There's still work to be done:

- [x] Add support for right-click operations on multiple selected layers.  
- [x] Limit context-menu options based on layers selected (eg 'rename' would only apply to single layer selections but 'zoom to extent' could take many layers). 
- [ ] Write tests
- [x] Remove redundant operations from left sidebar (likely everything under `'Geo Actions'`). 
- [x] Remove icons from layerlist layer item (the 'duplicate' operation should be moved to the right-click menu, the 'remove layer' operation should be removed). _Should this be done?_
- [ ] Add 'right-click' tip to "Welcome to dropchop!" text

I'm opening this up to get feedback.  Is this the right direction for DropChop?

alukach commented on src/js/menus.layer-context-menu.dropchop.js:
```diff
@@ -0,0 +1,89 @@
+      'expand': dc.ops.file['expand'],
+      'combine': dc.ops.file['combine'],
+      'rename': dc.ops.file['rename'],
+      'remove': dc.ops.file['remove'],
```
This was written before I saw how #208 rewrote how functions were loaded with the Left Sidebar Menu.  This could probably be refactored to more closely match that style for the sake of simplicity.  One win that this style has is that it could allow operations to be pulled from different files.

alukach commented:
Supporting multiple layers turned out to be easier than I expected.

![right-click-layer](https://cloud.githubusercontent.com/assets/897290/10721399/cf7a6a44-7b68-11e5-8c53-396893b43b49.gif)

mapsam commented:
Wowowow this is great! I'm very much into the idea of splitting menus into their own file, since they are growing larger. This makes a lot more sense (i.e. feels more *droppychoppy*) as well - the left-hand "geo actions" felt a bit forced, and very unintuitive since it all related to the layers you had selected anyways. :+1: for pushing this :smiley:

alukach commented:
@mapsam great, glad you're into it.  Feel free to add some styling changes/suggestions to the menu.

Got validation on the context-menu operations completed:

![right-click-layer](https://cloud.githubusercontent.com/assets/897290/10721828/434c581e-7b6f-11e5-995b-edf9c578863d.gif)

alukach commented:
Update on this:

* Removed icons on side of layers, add an ellipses to indicate "more options". Left-clicking on ellipses button triggers right-click event (for touch-screen users)
* Added headers to the context-menu to allow the operations to be organized. Not sure about this feature, thoughts?
* **problem:** Currently, it's possible for a context menu to go offscreen.  This can occur when right-clicking on layers near the bottom of the layerlist. Not sure what to do about this, adding a scrollbar (and autosizing the context menu) seems difficult.

![right-click-menu](https://cloud.githubusercontent.com/assets/897290/10875194/c2d7eaa8-80ec-11e5-852a-bd81aae66e35.gif)

mapsam commented:
This is looking really great.

I think the overflow issue can be solved by doing a little bit of math - checking the height of the menu and the window and ensuring it never generates beyond it. I'd be happy to work on this part sometime this week!

alukach commented:
@mapsam feel free to take a crack on keeping the menu on the screen. I admit that I hadn't really thought about just comparing the `div` size and placement to the window size and to ensure that it stays on screen, that sounds like a solid solution.

mapsam commented:
I'll take a look - I'm in my last week out in Richmond for Code for America so might not get to it this week - feel free to try and merge in and I can take a crack in a separate branch!

alukach commented:
![right-click-menu2](https://cloud.githubusercontent.com/assets/897290/11014462/954706ae-84f3-11e5-9144-141e62682001.gif)

Some style changes. Just to add some tests now...

mapsam commented:
Added the portion to calculate the menu location based on the window size. Seems to be working nicely. I'm a big fan of this menu system, it's much improved @alukach!

![dropchop-contextmenuoverflow](https://cloud.githubusercontent.com/assets/1943001/11015991/934fb14a-8541-11e5-932b-966d17a7663d.gif)

mapsam commented:
@alukach whenever you think this is ready to roll, feel free to merge!
